src/extensions/checkProjectBilling.ts

```diff
@@ -0,0 +1,130 @@
+import * as clc from "cli-color";
+import * as opn from "open";
+
+import * as cloudbilling from "../gcp/cloudbilling";
+import { FirebaseError } from "../error";
+import { logger } from "../logger";
+import { logPrefix } from "./extensionsHelper";
+import * as prompt from "../prompt";
+import * as utils from "../utils";
+
+/**
+ * Logs to console if setting up billing was successful.
+ */
+function logBillingStatus(enabled: boolean, projectId: string): void {
+  if (!enabled) {
+    throw new FirebaseError(
+      `${logPrefix}: ${clc.bold(
+        projectId
+      )} could not be upgraded. Please add a billing account via the Firebase console before proceeding.`
+    );
+  }
+  utils.logLabeledSuccess(logPrefix, `${clc.bold(projectId)} has successfully been upgraded.`);
+}
+
+/**
+ * Opens URL if applicable and stalls until user responds.
+ */
+async function openBillingAccount(projectId: string, url: string, open: boolean): Promise<boolean> {
+  if (open) {
+    try {
+      opn(url);
+    } catch (err) {
+      logger.debug("Unable to open billing URL: " + err.stack);
+    }
+  }
+
+  await prompt.promptOnce({
+    name: "continue",
+    type: "confirm",
+    message:
+      "Press enter when finished upgrading your project to continue setting up your extension.",
+    default: true,
+  });
+  return cloudbilling.checkBillingEnabled(projectId);
+}
+
+/**
+ * Question prompts user to select billing account for project.
+ */
+async function chooseBillingAccount(
+  projectId: string,
+  extensionName: string,
+  accounts: cloudbilling.BillingAccount[]
+): Promise<void> {
+  const choices = accounts.map((m) => m.displayName);
+  choices.push("Add new billing account");
+
+  const answer = await prompt.promptOnce({
+    name: "billing",
+    type: "list",
+    message: `The extension ${clc.underline(
+      extensionName
+    )} requires your project to be upgraded to the Blaze plan. You have access to the following billing accounts.
+Please select the one that you would like to associate with this project:`,
+    choices: choices,
+  });
+
+  let billingEnabled: boolean;
+  if (answer === "Add new billing account") {
+    const billingURL = `https://console.cloud.google.com/billing/linkedaccount?project=${projectId}`;
+    billingEnabled = await openBillingAccount(projectId, billingURL, true);
+  } else {
+    const billingAccount = accounts.find((a) => a.displayName == answer);
+    billingEnabled = await cloudbilling.setBillingAccount(projectId, billingAccount!.name);
+  }
+
+  return logBillingStatus(billingEnabled, projectId);
+}
+
+/**
+ * Directs user to set up billing account over the web and stalls until
+ * user responds.
+ */
+async function setUpBillingAccount(projectId: string, extensionName: string) {
+  const billingURL = `https://console.cloud.google.com/billing/linkedaccount?project=${projectId}`;
+
+  logger.info();
```

Comment:
Is this intentional to add a newline?

Reply:
Yep, this was in the old JS code. We use this pattern in a bunch of places throughout the CLI to space out logs/prompts to be more readable. Personally, I like it since it makes the code resemble what is going to be printed